add txt

dmytro-kulchytskyi opened the pull request:
Last version of proj

dmytro-kulchytskyi committed:
add txt

dmytro-kulchytskyi committed:
my changes

andrewyakonyuk commented on NewsWebSite/Views/Home/Article.cshtml:
```diff
@@ -0,0 +1,15 @@
+
+<div class="news_view">
+    <h2 class="news_head">@Model.Title</h2>
+    <a href="http://localhost:57821/Home/EditArticle?Id=@Model.Id">EDIT</a>
```
Используй UrlHelper. Для примера: Url.Action("EditArticle", "Home", new { id = Model.Id })

dmytro-kulchytskyi replied:
да, я юзаю его обычно, но у меня браузер видит его как home/home/SomeMethod.. 
поэтому так написал

andrewyakonyuk commented on NewsWebSite/Views/Home/Article.cshtml:
```diff
@@ -0,0 +1,15 @@
+    <a href="http://localhost:57821/Home/EditArticle?Id=@Model.Id">EDIT</a>
+    <img class="FullItemImg" src="~/Content/UserImages/@Model.Id/@Model.Image" alt="SRC" style="max-height: 400px">
+    <p>@Model.FullDescription</p>
+    <div class="publicate_time">@Html.Raw(createmsg)</div>
```
Крайне хреново. 
Для этого есть форматирование дат. 
[Статья в тему](http://metanit.com/sharp/tutorial/19.2.php)

dmytro-kulchytskyi committed:
добавил проверку на то, была ли изменена картинка(по весу)

andrewyakonyuk commented on NewsWebSite/Views/Home/CreateArticle.cshtml:
```diff
@@ -0,0 +1,74 @@
+        @Html.LabelFor(model => Model.Image)
+        @Html.TextBoxFor(model => model.Image, new { type = "file", accept = "image/jpeg,image/png,.jpg,.png", @id="inputfile"})
+        @Html.ValidationMessageFor(model => model.Image)
+    </div>
```
Почти классно, но в этом случае можно сделать проще:
Html.EditorForModel() - он создаст весь html, основываясь на аттрибутах(required, display etc).
Например
`@using (Html.BeginForm("CreateArticle", "Home", FormMethod.Post, new { enctype = "multipart/form-data" }))
{
    Html.EditorForModel()
}
`

dmytro-kulchytskyi replied:
да, я знаю про такую штуку, но если делать так, то для файла оно неправильный инпут создает, а так есть возможность дописать new {type = "file"} и оно все нормально отображает

andrewyakonyuk commented on NewsWebSite/Attributes/ValidImage.cs:
```diff
@@ -0,0 +1,37 @@
+        {
+            //return false;
+            if (value == null) return false;
+            int maxContent = 5 * 1024 * 1024; //3 MB
```
Использование аттрибута для валидации одобряю.

Теперь о плохом:
1. `5 * 1024 * 1024` это ниразу не 3mb. Коммент говно. И другие такие же поудаляй
2. Максимальный размер файла захардкожен. Поэтому, нужно позволить считывать его, например, из свойства MaxFileSize. В таком случае, если тебе понадобится провалидировать размер фото, тебе не нужно менять сам аттрибут. 
3. Реальный максимальный размер файла, который можно загрузить на сервер, на самом деле деле [конфигурируется отдельно в web.config](http://stackoverflow.com/questions/288612/how-to-increase-the-max-upload-file-size-in-asp-net). 
4. Это аттрибут делает слишком много: он валидирует и размер файла, и его тип. Разбей на два отдельный. 
5. Размер файла в аттрибуте и в яваскрипте забиты отдельно и ни как не связаны. К примеру, если нужно поменять с 3(?)мб до 100, тебе прийдется искать все места, где он используется.
6. И вообще, используй [jquery unobtrusive validation](http://stackoverflow.com/questions/11534910/what-is-jquery-unobtrusive-validation), [на русском](https://habrahabr.ru/post/180279/)

andrewyakonyuk commented on NewsWebSite/Attributes/ValidImageEditArticle.cs:
```diff
@@ -0,0 +1,37 @@
+
+namespace NewsWebSite.Attributes
+{
+    public class ValidImageEditArticle : ValidationAttribute
```
Зачем два одинаковый аттрибута?

andrewyakonyuk commented on NewsWebSite/Controllers/HomeController.cs:
```diff
@@ -30,148 +38,101 @@ public ActionResult CreateLines(int n = 0)
+            if (edited.Image != null)
+            {
+                FIleHelper.SaveOrUpdateArticleImage(Server.MapPath(ConfigurationManager.AppSettings["UserImagesFolder"].ToString()), edited.Image, baseArticle.Id);
+                baseArticle.Image = edited.Image.FileName;
```
Использование `ConfigurationManager` одобряю.

Имеет смысл использовать путь к папке с картинками только в `FileHelper`.
`ConfigurationManager.AppSettings["UserImagesFolder"].ToString()` - ToString() не нужен

andrewyakonyuk commented on NewsWebSite/Controllers/HomeController.cs:
```diff
@@ -30,148 +38,101 @@ public ActionResult CreateLines(int n = 0)
+            }
+            if (edited.FullDescription != null)
+            {
+                baseArticle.FullDescription = edited.FullDescription;
```
проверки вида `edited.FullDescription != null` не нужны. Для этого уже есть валидационные аттрибуты.

dmytro-kulchytskyi replied:
задумка в том, что если юзер оставит пустое поле, то значение останется старое и ошибки не будет, в атрибутах для editArticle нету [required], и они пропускают null'ы

andrewyakonyuk commented on NewsWebSite/Controllers/HomeController.cs:
```diff
@@ -30,148 +38,101 @@ public ActionResult CreateLines(int n = 0)
-
-        public ActionResult About()
+        [HttpPost]
+        public ActionResult GetNewPageOfArticles(int Page)
```
параметры методов и переменный всегда начинаются с маленькой буквы.
Смотри в [naming guidelines](http://stackoverflow.com/a/8497191)

andrewyakonyuk commented on NewsWebSite/Models/Article.cs:
```diff
@@ -12,10 +15,17 @@ public class Article
-        public virtual DateTime LastUpdateDate { get; set; }
-
+        public virtual string CreateDate { get; set; }
+        public virtual string LastUpdateDate { get; set; }
```
Даты остаются с типом `DateTime`. Так что вертай как было.
Для форматирования есть другие средства + есть такая штука, как локализация, то бишь не для всех языков порядок день-месяц-год один и тот же. В Америке принят  месяц-день-год. Учитывай это

andrewyakonyuk commented on NewsWebSite/Models/Article.cs:
```diff
@@ -12,10 +15,17 @@ public class Article
 
+        public Article(EditArticleModel a)
+        {
+            Title = a.Title;
```
Обьект, который сохраняется в базу, ничего не должен знать об обьектах, которые используются для отображения. Это как минимум потому, что у тебя может быть одна база, но несколько приложений, которые ее используют (к примеру веб и mobile apps).
Гугли [трехуровневую архитектуру](https://ru.wikipedia.org/wiki/%D0%A2%D1%80%D1%91%D1%85%D1%83%D1%80%D0%BE%D0%B2%D0%BD%D0%B5%D0%B2%D0%B0%D1%8F_%D0%B0%D1%80%D1%85%D0%B8%D1%82%D0%B5%D0%BA%D1%82%D1%83%D1%80%D0%B0). [Более подробно](https://msdn.microsoft.com/en-us/library/ee658109.aspx)

andrewyakonyuk commented on NewsWebSite/Models/IRepository.cs:
```diff
@@ -12,111 +14,144 @@ public interface IRepository
+        IList<Article> GetList(int starFrom, int count);
+        PagedList<DemoArticle> GetDemoList(int starFrom, int count);
+        int SaveOrUpdate(Article article);
+        int SaveOrUpdate(CreateArticleModel model);
```
Как по мне, так просто `Save` намного лучше смотрится, так как он и предролагает и создание, и обновление.

andrewyakonyuk commented on NewsWebSite/Models/IRepository.cs:
```diff
@@ -12,111 +14,144 @@ public interface IRepository
+                using (var t = session.BeginTransaction())
+                {
+                    var timeNow = DateTime.Now.ToString("MM/dd/yy H:mm"); ;
+                    if (a.CreateDate == null)
```
Вернуть тип `DateTime`. Тогда проверка будет выглядеть `if(a.CreateDate == DateTime.MinValue)`

andrewyakonyuk commented on NewsWebSite/Models/IRepository.cs:
```diff
@@ -12,111 +14,144 @@ public interface IRepository
+       
+        public int SaveOrUpdate(CreateArticleModel model)
+        {
+            Article a = new Article();
```
[Репозиторий](http://metanit.com/sharp/articles/mvc/11.php) ничего не должен знать о таких моделях. Он работает только с `Article`

andrewyakonyuk commented on NewsWebSite/Models/IRepository.cs:
```diff
@@ -12,111 +14,144 @@ public interface IRepository
-        public IList<Article> GetList(int starfrom = 0, int count = 10)
+
+        public IList<Article> GetList(int starFrom = 0, int count = 10)
         {
```
В этом методе не хватает `results.TotalCount = session.QueryOver<Article>().Select(Projections.RowCount()).FutureValue<int>().Value;`. 
Одинаковые вещи должны работать одинаково.

andrewyakonyuk commented on NewsWebSite/Models/IRepository.cs:
```diff
@@ -12,111 +14,144 @@ public interface IRepository
             }
         }
+
+        public string GetDemoListJson(int startFrom, int count)
```
Методу здесь не место. Перенеси его в контроллер. Репозиторий только сохраняет и выполняет выборку данных.

andrewyakonyuk commented on NewsWebSite/Models/IRepository.cs:
```diff
@@ -12,111 +14,144 @@ public interface IRepository
+            return JsonConvert.SerializeObject(list);
+        }
+
         public int GetCountOfLines()
```
Его можно уже удалить

andrewyakonyuk commented on NewsWebSite/Controllers/HomeController.cs:
```diff
@@ -30,148 +38,101 @@ public ActionResult CreateLines(int n = 0)
 
-        public ActionResult Contact()
+        [HttpPost]
+        public ActionResult GetNPagesOfArticles(int n)
```
В чем принципиальная разница между `GetNPagesOfArticles` и `GetNewPageOfArticles`?
Соединить в один метод.
Даже больше: тебе ну нужен отдельный метод для рендеринга новостей и для lazy loading. Можно обойтись одним методом:
```
public ActionResult Index(int page = 0)
{
        var list = repo.GetDemoList(page, NumberOfItemsOnPage);
        if (Request.IsAjaxRequest())
        {
            return Json(list);
        }
        var model = new ListItemPageModel(NumberOfItemsOnPage, list);
        return View(model);
}
```

andrewyakonyuk commented on NewsWebSite/Models/ViewModel/ListItemPageModel.cs:
```diff
@@ -0,0 +1,19 @@
+        public int NumberOfItemsOnPage { get; set; }
+        public PagedList<DemoArticle> ArticleList { get; set; }
+
+        public ListItemPageModel(int NumberOfItemsOnPage, PagedList<DemoArticle> ArticleList)
```
Имена, блеать!

andrewyakonyuk commented on NewsWebSite/Views/Home/CreateArticle.cshtml:
```diff
@@ -0,0 +1,74 @@
+
+        var _validFileExtensions = [".jpg", ".jpeg", ".bmp", ".gif", ".png"];
+        function Validate(oForm) {
+            var arrInputs = $("input[for='Image']").val();
```
скрипт говно. гугли jquery validation

andrewyakonyuk reviewed (changes requested):
Пару пунктов на доработку

dmytro-kulchytskyi committed:
Учел все, что ты написал

@html.display4model я пробывал, но он не умеет работать с файловами
инпутами, так же как и @html.dilplay4.
я это обошел так
@Html.TextBoxFor(model => model.Image, new { type = "file", accept =
"..." })

dmytro-kulchytskyi commented:
почти все, пару пунктов не заметил(были свернутые)
P.S. DATETIME вернул еще в коммите с проверкой на изменение картинки, и в этом же коммите заменил все стчные ссылки на @url.action
забыл в описании указать это

dmytro-kulchytskyi committed:
теперь точно все)

dmytro-kulchytskyi commented:
не получается вернуть json(return Json(lst)), jquery выдает ошибку что не может его прочитать на 1 символе, хотя в консольке хрома он показывается как нормлаьный json обьект.
Минус метода который я использую в том, что он возвращает дату как строку, и ее форматировать сложно

dmytro-kulchytskyi committed:
remove nhib cfg to web cfg

andrewyakonyuk commented on NewsWebSite/Models/Repository/IRepository.cs:
```diff
@@ -0,0 +1,150 @@
+        int Save(Article article);
+    }
+
+    //public class NHibernateRepository : IRepository
```
удали весь этот коммент

andrewyakonyuk commented on NewsWebSite/Web.config:
```diff
@@ -5,14 +5,32 @@
+  </hibernate-configuration>
+
   <connectionStrings>
     <add name="DefaultConnection" connectionString="Data Source=(LocalDb)\MSSQLLocalDB;AttachDbFilename=|DataDirectory|\aspnet-NewsWebSite-20170115063838.mdf;Initial Catalog=aspnet-NewsWebSite-20170115063838;Integrated Security=True" providerName="System.Data.SqlClient" />
```
и это тоже удали

andrewyakonyuk commented on NewsWebSite/Web.config:
```diff
@@ -5,14 +5,32 @@
+        NHibernate.Driver.SqlClientDriver
+      </property>
+      <property name="connection.connection_string">
+        Server=DIMAS\SQLEXPRESS; database=NewsWebSiteDB; Integrated Security=False; User Id=root; Password=root; MultipleActiveResultSets=True;
```
это `Server=DIMAS\SQLEXPRESS` будет работать только на твоем ноуте. Замени на ` Server=.\SQLEXPRESS`

dmytro-kulchytskyi committed:
fix
